pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1791,7 +1792,30 @@ public int messagesCount() {
 
         @Override
         public Iterator<OpSendMsg> iterator() {
-            return delegate.iterator();
+            Iterator<OpSendMsg> delegateIterator = delegate.iterator();
+            return new Iterator<OpSendMsg>() {
+
+                @Override
+                public boolean hasNext() {
+                    return delegateIterator.hasNext();
+                }
+
+                @Override
+                public OpSendMsg next() {
+                    return delegateIterator.next();
+                }
+
+                @Override
+                public void remove() {
+                    delegateIterator.remove();
+                    messagesCount.addAndGet(-1);
+                }
```

Comment:
It seems that `numMessagesInBatch` would have to be taken into account too. Something like this:
```suggestion
                OpSendMsg currentOp;
                
                @Override
                public boolean hasNext() {
                    return delegateIterator.hasNext();
                }

                @Override
                public OpSendMsg next() {
                    currentOp = delegateIterator.next();
                    return currentOp;
                }

                @Override
                public void remove() {
                    delegateIterator.remove();
                    if (currentOp != null) {
                       messagesCount.addAndGet(-currentOp.numMessagesInBatch);
                       currentOp = null;
                     }
                }
```